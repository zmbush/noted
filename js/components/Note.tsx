// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as ReactMarkdown from 'react-markdown';
import classNames from 'classnames';

import axios from 'axios';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardHeader from '@material-ui/core/CardHeader';
import CardActions from '@material-ui/core/CardActions';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import EditIcon from '@material-ui/icons/Edit';
import SaveIcon from '@material-ui/icons/Save';
import CreateNewFolderIcon from '@material-ui/icons/CreateNewFolder';
import DeleteIcon from '@material-ui/icons/Delete';
import ChipInput from 'material-ui-chip-input';
import InputBase from '@material-ui/core/InputBase';
import {
  createStyles,
  withStyles,
  Theme,
  WithStyles,
} from '@material-ui/core/styles';

import 'codemirror/lib/codemirror.css';
import 'tui-editor/dist/tui-editor.min.css';
//import 'tui-editor/dist/tui-editor-contents.min.css';
import { Editor } from '@toast-ui/react-editor';

import { NoteData } from 'data/types';
import BindKeyboard from 'components/BindKeyboard';
import Tags from 'components/Tags';
import AutoLink from 'components/AutoLink';

const styles = (theme: Theme) =>
  createStyles({
    bodyRoot: {
      width: '100%',
    },
    card: {
      '@media print': {
        border: 'none',
        boxShadow: 'none',
        //pageBreakInside: 'avoid',
      },
    },
    unlinked: {
      border: '5px solid red',
    },
    bodyEditor: {
      fontFamily: 'Roboto Mono',
    },
    markdown: {
      '& p': {
        '@media print': {
          //pageBreakInside: 'avoid',
          //display: 'inline-block',
          //margin: 0,
          textIndent: theme.spacing.unit,
        },
      },
      '& blockquote': {
        borderLeftWidth: 3,
        borderLeftStyle: 'solid',
        borderLeftColor: theme.palette.secondary.dark,

        paddingLeft: theme.spacing.unit,
        '@media print': {
          margin: 0,
        },
      },
      '& a': {
        '@media print': {
          textDecoration: 'none',
          color: 'black',
        },
      },
      '& table': {
        borderCollapse: 'collapse',
        borderSpacing: 0,
        width: '100%',

        '& thead': {
          fontWeight: 800,
          display: 'table-row-group',

          '& th': {
            verticalAlign: 'bottom',
            paddingBottom: '.3em',
            paddingLeft: '.1em',
            paddingRight: '.1em',
          },
        },
        '& tbody': {
          '& tr:nth-child(2n+1)': {
            backgroundColor: '#E0E5C1',
          },
        },
      },
      '@media print': {
        '& h1, & h2, & h3, & h4, & h5': {
          color: '#58180D',
        },
        '& h1': {
          fontSize: '.705cm',
        },
        '& h2': {
          fontSize: '.529cm',
          borderBottom: '2px solid #c9ad6a',
        },
        '& h3': {
          fontSize: '.458cm',
          marginBottom: 0,
        },
        '& h4': {
          fontSize: '.423cm',
          marginBottom: '.2em',
        },
      },
    },
    noPrint: {
      '@media print': {
        display: 'none',
      },
    },
    cardHeader: {
      '@media print': {
        paddingBottom: 0,
        paddingTop: 0,
        '& span': {
          color: '#58180D',
          fontSize: '.987cm',
        },
      },
    },
    cardContent: {
      '@media print': {
        paddingTop: 0,
      },
    },
  });

const editorToolbarConfig = {
  options: ['inline', 'blockType', 'list', 'link', 'remove', 'history'],
  inline: {
    options: ['bold', 'italic', 'strikethrough'],
  },
};

const draftMarkdownOptions = {
  styleItems: {
    'color-red': {
      open: () => '<span class="style-red">',
      close: () => '</span>',
    },
  },
  remarkablePreset: 'full',
  remarkableOptions: { html: true, linkify: true, typographer: true },
};

/*const toMarkdown = (editorState: EditorState) =>
  draftToMarkdown(
    convertToRaw(editorState.getCurrentContent()),
    draftMarkdownOptions
  );

const toDraft = (markdown: string) =>
  EditorState.createWithContent(
    convertFromRaw(markdownToDraft(markdown, draftMarkdownOptions))
    );*/

interface Props extends WithStyles<typeof styles> {
  new?: boolean;
  note: NoteData;
  titles: Map<string, Set<number>>;
  updateNote: (note?: { id: number }) => void;
  matches?: {
    indices: number[][];
    value: string;
    key: string;
    arrayIndex: number;
  }[];
}

const initialState = {
  edit: false,
  title: '',
  body: '',
  tags: new Array<String>(),
};

type State = Readonly<typeof initialState>;

class Note extends React.Component<Props, State> {
  mainInput: React.RefObject<HTMLInputElement>;
  editor: React.RefObject<Editor>;

  constructor(props: Props) {
    super(props);
    this.state = initialState;
    this.mainInput = React.createRef();
    this.editor = React.createRef();
  }

  cancelEdit = (e: React.SyntheticEvent | Event) => {
    e.preventDefault();
    if (this.state.body == this.props.note.body) {
      this.setState({ edit: false });
      this.props.updateNote(null);
    }
  };

  saveShortcut = (e: React.SyntheticEvent | Event) => {
    e.preventDefault();
    this.save();
  };

  save = async () => {
    let result;
    if (this.props.new) {
      result = await axios.put('/api/secure/note', {
        title: this.state.title,
        body: this.state.body,
      });
    } else {
      result = await axios.patch(`/api/secure/notes/${this.props.note.id}`, {
        title: this.state.title,
        body: this.state.body,
      });
    }

    result = await axios.put(
      `/api/secure/notes/${result.data.id}/tags`,
      this.state.tags
    );

    this.setState({
      edit: false,
    });

    this.props.updateNote(result.data);
  };

  startEdit = () => {
    this.setState(
      {
        title: this.props.note.title,
        body: this.props.note.body,
        tags: this.props.note.tags,
        edit: true,
      },
      () => {
        let e = this.mainInput.current;
        e.focus();
        e.setSelectionRange(e.value.length, e.value.length);
      }
    );
  };

  addTag = (tag: string) => {
    this.setState({
      tags: [...this.state.tags, tag],
    });
  };

  deleteTag = (tag: string, index: number) => {
    this.state.tags.splice(index, 1);
    this.setState({
      tags: this.state.tags,
    });
  };

  render() {
    const { classes } = this.props;

    return (
      <BindKeyboard keys='ctrl+s' callback={this.saveShortcut}>
        <BindKeyboard keys='esc' callback={this.cancelEdit}>
          <Card
            className={classNames(classes.card, {
              [classes.unlinked]: this.props.note.user_id == 1,
            })}
          >
            <CardHeader
              className={classes.cardHeader}
              title={
                this.state.edit ? (
                  <InputBase
                    value={this.state.title}
                    onChange={e => {
                      this.setState({ title: e.target.value });
                    }}
                    style={{ fontSize: 'inherit' }}
                  />
                ) : (
                  this.props.note.title
                )
              }
              action={
                this.state.edit ? (
                  <IconButton onClick={this.save}>
                    <SaveIcon />
                  </IconButton>
                ) : (
                  <IconButton
                    onClick={this.startEdit}
                    className={classes.noPrint}
                  >
                    <EditIcon />
                  </IconButton>
                )
              }
            />
            <CardContent className={classes.cardContent}>
              {this.state.edit ? (
                <>
                  <ChipInput
                    classes={{}}
                    placeholder='Tags'
                    fullWidth
                    dataSource={[
                      'arc:Delmirev',
                      'type:Location',
                      'type:Character',
                    ]}
                    value={this.state.tags}
                    onAdd={this.addTag}
                    onDelete={this.deleteTag}
                  />
                  <Editor
                    initialValue={this.state.body}
                    initialEditType='wysiwyg'
                    ref={this.editor}
                    onChange={() => {
                      if (this.editor.current) {
                        this.setState({
                          body: this.editor.current.editorInst.getMarkdown(),
                        });
                      }
                    }}
                    height='auto'
                    usageStatistics={false}
                  />
                  <div style={{ display: 'none' }}>
                    <InputBase
                      inputRef={this.mainInput}
                      value={this.state.body}
                      onChange={e => {
                        this.setState({
                          body: e.target.value,
                        });
                      }}
                      style={{ fontSize: 'inherit' }}
                      classes={{
                        root: classes.bodyRoot,
                        multiline: classes.bodyEditor,
                      }}
                      multiline
                      rows={15}
                    />
                  </div>
                </>
              ) : (
                <>
                  <Tags tags={this.props.note.tags} />
                  <ReactMarkdown
                    className={classes.markdown}
                    renderers={{
                      text: props => (
                        <AutoLink titles={this.props.titles} {...props} />
                      ),
                    }}
                  >
                    {this.props.note.body}
                  </ReactMarkdown>
                </>
              )}
            </CardContent>
          </Card>
        </BindKeyboard>
      </BindKeyboard>
    );
  }
}

export type InnerNote = Note;

export default withStyles(styles)(Note);
