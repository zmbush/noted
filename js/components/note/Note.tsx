// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import rehypeRaw from 'rehype-raw';

import * as React from 'react';
import { Suspense } from 'react';
import ReactLoading from 'react-loading';
import ReactMarkdown from 'react-markdown';
import { ReactMarkdownOptions } from 'react-markdown/lib/react-markdown';
import { useDispatch, useSelector } from 'react-redux';

import {
  Archive as ArchiveIcon,
  Clear as UnpinIcon,
  Delete as DeleteIcon,
  Done as PinIcon,
  Edit as EditIcon,
  LibraryAdd as LibraryAddIcon,
  MoreVert as MoreVertIcon,
  Unarchive as UnarchiveIcon,
} from '@mui/icons-material';
import {
  Card,
  CardContent,
  CardHeader,
  Dialog,
  Grid,
  IconButton,
  ListItemIcon,
  Menu,
  MenuItem,
  useMediaQuery,
  useTheme,
} from '@mui/material';

import ConfirmationDialog from 'components/core/ConfirmationDialog';
import Loading from 'components/core/Loading';
import AutoLink from 'components/note/AutoLink';
import Tags from 'components/note/Tags';
import { createNote, deleteNote, updateNote } from 'data/notes/api';
import { getLinkIds } from 'data/notes/selectors';
import { AppState } from 'data/store';
import { NewNote, UpdateNote, NoteWithTags } from 'data/types';
import { getIsNoteChanging } from 'data/ui/selectors';

import * as styles from './styles.scss';

const NoteEditor = React.lazy(
  () => import(/* webpackChunkName: "editor" */ 'components/note/NoteEditor'),
);

type NoteContentsProps = {
  note: NoteWithTags;
  titles: { [title: string]: Set<number> };
  setEdit: (edit: boolean) => void;
  setCreatingSubNote: (creatingSubNote: boolean) => void;
  children?: React.ReactElement | React.ReactElement[];
};

export const NoteContents = ({
  note,
  setEdit,
  setCreatingSubNote,
  titles,
  children,
}: NoteContentsProps) => {
  const [moreMenuEl, setMoreMenuEl] = React.useState<HTMLElement | null>(null);
  const [confirmDeleteOpen, setConfirmDeleteOpen] = React.useState(false);
  const dispatch = useDispatch();
  const noteChanging = useSelector((state: AppState) => getIsNoteChanging(state, note));

  const doDelete = () => dispatch(deleteNote(note.id));

  const archiveNote = async () => {
    setMoreMenuEl(null);

    const { id, archived } = note;
    await dispatch(updateNote({ id, archived: !archived }));
  };

  const pinNote = async () => {
    setMoreMenuEl(null);

    const { id, pinned } = note;
    await dispatch(updateNote({ id, pinned: !pinned }));
  };

  const startEdit = () => {
    setEdit(true);
  };

  const startSubNoteCreate = () => {
    setCreatingSubNote(true);
  };

  const markdownComponents: ReactMarkdownOptions['components'] = {
    // eslint-disable-next-line react/no-unstable-nested-components
    p: ({ children: pChildren }) => (
      <p>
        <AutoLink titles={titles}>{pChildren}</AutoLink>
      </p>
    ),
  };

  return (
    <Card
      raised
      sx={[
        {
          marginBottom: 1,
          position: 'relative',
          '@media print': {
            border: 'none',
            boxShadow: 'none',
            // pageBreakInside: 'avoid',
          },
        },
        note.user_id === 1
          ? {
              border: '5px',
              borderColor: 'red',
            }
          : {},
        note.archived
          ? {
              opacity: 0.4,
              '@media print': {
                display: 'none',
              },
            }
          : {},
      ]}
    >
      <CardHeader
        sx={{
          '@media print': {
            paddingBottom: 0,
            paddingTop: 0,
            '& span': {
              color: '#58180d',
              fontSize: '0.987cm',
            },
          },
        }}
        avatar={note.pinned ? <PinIcon /> : null}
        title={note.title}
        action={
          <>
            <IconButton
              onClick={startSubNoteCreate}
              sx={{ displayPrint: 'none' }}
              aria-label='Add SubNote'
              size='large'
            >
              <LibraryAddIcon />
            </IconButton>
            <IconButton
              onClick={startEdit}
              sx={{ displayPrint: 'none' }}
              aria-label='Edit Note'
              size='large'
            >
              <EditIcon />
            </IconButton>
            <IconButton
              onClick={(e) => setMoreMenuEl(e.currentTarget)}
              sx={{ displayPrint: 'none' }}
              aria-owns={moreMenuEl ? 'more-menu' : undefined}
              aria-label='More Options'
              size='large'
            >
              <MoreVertIcon />
            </IconButton>
          </>
        }
      />
      <Menu
        id='more-menu'
        anchorEl={moreMenuEl}
        open={Boolean(moreMenuEl)}
        onClose={() => setMoreMenuEl(null)}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'left' }}
        transformOrigin={{ vertical: 'top', horizontal: 'left' }}
      >
        <MenuItem onClick={() => setConfirmDeleteOpen(true)}>
          <ListItemIcon>
            <DeleteIcon />
          </ListItemIcon>
          Delete Note
        </MenuItem>
        <MenuItem onClick={pinNote}>
          <ListItemIcon>{note.pinned ? <UnpinIcon /> : <PinIcon />}</ListItemIcon>
          {note.pinned ? 'Unpin Note' : 'Pin Note'}
        </MenuItem>
        <MenuItem onClick={archiveNote}>
          <ListItemIcon>{note.archived ? <UnarchiveIcon /> : <ArchiveIcon />}</ListItemIcon>
          {note.archived ? 'Unarchive Note' : 'Archive Note'}
        </MenuItem>
      </Menu>
      <ConfirmationDialog
        open={confirmDeleteOpen}
        title={`You are about to delete note: ${note.title}`}
        onPositive={doDelete}
        onNegative={() => setConfirmDeleteOpen(false)}
      />
      <Loading
        in={noteChanging}
        timeout={1000}
        sx={{
          position: 'absolute',
          margin: 'auto',
          width: 80,
          height: 80,
          left: 0,
          right: 0,
          top: 40,
        }}
      />
      <CardContent
        sx={{
          paddingTop: 0,
          '@media print': {
            paddingLeft: 0,
            '&:last-child': {
              paddingBottom: 0,
            },
          },
        }}
      >
        <Tags tags={note.tags} />
        <ReactMarkdown
          className={styles.markdown}
          components={markdownComponents}
          rehypePlugins={[rehypeRaw]}
        >
          {note.body}
        </ReactMarkdown>
        <Grid
          container
          sx={{
            marginTop: 4,
            '@media print': {
              marginTop: 0,
              display: 'block',
            },
          }}
        >
          {children}
        </Grid>
      </CardContent>
    </Card>
  );
};

type Props = {
  children?: React.ReactElement | React.ReactElement[];
} & (
  | { note: NewNote; onNewNoteCancel: () => void }
  | { note: NoteWithTags; onNewNoteCancel?: never }
);

const Note = ({ note, onNewNoteCancel, children }: Props) => {
  const [edit, setEdit] = React.useState(!('id' in note)); // If note has no 'id', it must be a NewNote
  const [creatingSubNote, setCreatingSubNote] = React.useState(false);
  const [confirmCancelEditOpen, setConfirmCancelEditOpen] = React.useState(false);
  const [noteIsModified, setNoteIsModified] = React.useState(false);
  const theme = useTheme();
  const editorFullscreen = useMediaQuery(theme.breakpoints.down('xs'));
  const dispatch = useDispatch();
  const titles = useSelector(getLinkIds);

  const cancelEdit = () => {
    setEdit(false);
    setCreatingSubNote(false);
    setConfirmCancelEditOpen(false);
    if (onNewNoteCancel) {
      onNewNoteCancel();
    }
  };

  const tryCancelEdit = () => {
    if (noteIsModified) {
      setConfirmCancelEditOpen(true);
    } else {
      cancelEdit();
    }
  };

  const save = async (noteData: (NewNote | UpdateNote) & Pick<NoteWithTags, 'tags'>) => {
    if (creatingSubNote || !('id' in note)) {
      await dispatch(createNote(noteData as NewNote & Pick<NoteWithTags, 'tags'>));
    } else {
      await dispatch(
        updateNote({
          id: note.id,
          ...noteData,
        }),
      );
    }

    setEdit(false);
    setCreatingSubNote(false);
  };

  return (
    <>
      {'id' in note ? (
        <NoteContents
          titles={titles}
          setEdit={setEdit}
          setCreatingSubNote={setCreatingSubNote}
          note={note}
        >
          {children}
        </NoteContents>
      ) : null}
      <ConfirmationDialog
        open={confirmCancelEditOpen}
        title='If you close this editor, you will lose your changes.'
        onPositive={cancelEdit}
        onNegative={() => setConfirmCancelEditOpen(false)}
      />
      <Dialog
        classes={{ root: styles.markdown }}
        open={edit || creatingSubNote}
        fullWidth
        maxWidth='lg'
        fullScreen={editorFullscreen}
        onClose={tryCancelEdit}
        data-testid='edit-note-dialog'
      >
        <Suspense
          fallback={<ReactLoading type='spin' className={styles.loadingSpinner} color='#000000' />}
        >
          {edit || creatingSubNote ? (
            <NoteEditor
              onSave={save}
              onModified={setNoteIsModified}
              note={
                'id' in note && creatingSubNote
                  ? { title: '', body: '', parent_note_id: note.id }
                  : note
              }
            />
          ) : null}
        </Suspense>
      </Dialog>
    </>
  );
};

export default Note;
