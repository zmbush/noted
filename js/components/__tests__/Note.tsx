// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
import { shallow } from 'enzyme';

import * as React from 'react';

import { Inner as Note, NoteContents } from '../Note';

const note = (
  <Note
    noteViewFilter={null}
    search=''
    titles={new Map()}
    subNotes={new Map()}
    onUpdateNote={() => {}}
    onDeleteNote={() => {}}
    note={{
      id: 1,
      title: 'note title',
      body: 'note body',
      tags: ['tag1'],
      created_at: '',
      updated_at: '',
      user_id: 2,
    }}
  />
);

const noteContents = (
  <NoteContents
    noteViewFilter={null}
    search=''
    titles={new Map()}
    subNotes={new Map()}
    onUpdateNote={() => {}}
    onDeleteNote={() => {}}
    setEdit={() => {}}
    setCreatingSubNote={() => {}}
    note={{
      id: 1,
      title: 'note title',
      body: 'note body',
      tags: ['tag1'],
      created_at: '',
      updated_at: '',
      user_id: 2,
      parent_note_id: 0,
      archived: false,
      pinned: false,
    }}
  />
);

describe('<Note />', () => {
  test('matches snapshot', () => {
    const wrapper = shallow(note);

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({
      note: {
        pinned: true,
      },
    });

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({
      note: {
        archived: true,
        pinned: false,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
});

describe('<NoteContents />', () => {
  test('matches snapshot', () => {
    const wrapper = shallow(noteContents);

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({
      note: {
        pinned: true,
      },
    });

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({
      note: {
        archived: true,
        pinned: false,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
});
