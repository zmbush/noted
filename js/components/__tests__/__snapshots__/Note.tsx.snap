// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Note /> matches snapshot 1`] = `
<WithStyles(Card)
  className=""
>
  <WithStyles(CardHeader)
    action={
      <React.Fragment>
        <WithStyles(IconButton)
          aria-label="Add Subnote"
          onClick={[Function]}
        >
          <pure(LibraryAddIcon) />
        </WithStyles(IconButton)>
        <WithStyles(IconButton)
          aria-label="Edit Note"
          onClick={[Function]}
        >
          <pure(EditIcon) />
        </WithStyles(IconButton)>
        <WithStyles(IconButton)
          aria-label="More Options"
          onClick={[Function]}
        >
          <pure(MoreVertIcon) />
        </WithStyles(IconButton)>
      </React.Fragment>
    }
    avatar={null}
    title="note title"
  />
  <WithStyles(Menu)
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "bottom",
      }
    }
    getContentAnchorEl={null}
    id="more-menu"
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
  >
    <WithStyles(MenuItem)
      onClick={[Function]}
    >
      <WithStyles(ListItemIcon)>
        <pure(DeleteIcon) />
      </WithStyles(ListItemIcon)>
      Delete Note
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      onClick={[Function]}
    >
      <WithStyles(ListItemIcon)>
        <pure(DoneIcon) />
      </WithStyles(ListItemIcon)>
      Pin Note
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      onClick={[Function]}
    >
      <WithStyles(ListItemIcon)>
        <pure(ArchiveIcon) />
      </WithStyles(ListItemIcon)>
      Archive Note
    </WithStyles(MenuItem)>
  </WithStyles(Menu)>
  <ConfirmationDialog
    message="Are you sure?"
    negative="No"
    onNegative={[Function]}
    onPositive={[Function]}
    open={false}
    positive="Yes"
    title="You are about to delete note: note title"
  />
  <ConfirmationDialog
    message="Are you sure?"
    negative="No"
    onNegative={[Function]}
    onPositive={[Function]}
    open={false}
    positive="Yes"
    title="If you close this editor, you will lose your changes."
  />
  <WithStyles(CardContent)>
    <WithStyles(Dialog)
      classes={
        Object {
          "root": undefined,
        }
      }
      fullScreen={false}
      fullWidth={true}
      maxWidth="lg"
      onClose={[Function]}
      open={false}
    >
      <React.Suspense
        fallback={
          <Loading
            color="#000000"
            delay={0}
            height={64}
            type="spin"
            width={64}
          />
        }
      >
        <lazy
          innerRef={
            Object {
              "current": null,
            }
          }
          note={
            Object {
              "body": "",
              "parent_note_id": 1,
              "tags": Array [],
              "title": "",
            }
          }
          onSave={[Function]}
        />
      </React.Suspense>
    </WithStyles(Dialog)>
    <WithStyles(Tags)
      tags={
        Array [
          "tag1",
        ]
      }
    />
    <ReactMarkdown
      astPlugins={
        Array [
          [Function],
        ]
      }
      escapeHtml={false}
      parserOptions={Object {}}
      plugins={Array []}
      rawSourcePos={false}
      renderers={
        Object {
          "text": [Function],
        }
      }
      skipHtml={false}
      sourcePos={false}
      transformLinkUri={[Function]}
    >
      note body
    </ReactMarkdown>
    <WithStyles(Grid)
      container={true}
      spacing={8}
    >
      <withRouter(Connect(WithStyles(NoteList)))
        deleteNote={[Function]}
        depth={1}
        notes={Map {}}
        parent_note_id={1}
        search=""
        updateNote={[Function]}
      />
    </WithStyles(Grid)>
  </WithStyles(CardContent)>
</WithStyles(Card)>
`;

exports[`<Note /> matches snapshot 2`] = `
<WithStyles(Card)
  className=""
>
  <WithStyles(CardHeader)
    action={
      <React.Fragment>
        <WithStyles(IconButton)
          aria-label="Add Subnote"
          onClick={[Function]}
        >
          <pure(LibraryAddIcon) />
        </WithStyles(IconButton)>
        <WithStyles(IconButton)
          aria-label="Edit Note"
          onClick={[Function]}
        >
          <pure(EditIcon) />
        </WithStyles(IconButton)>
        <WithStyles(IconButton)
          aria-label="More Options"
          onClick={[Function]}
        >
          <pure(MoreVertIcon) />
        </WithStyles(IconButton)>
      </React.Fragment>
    }
    avatar={<pure(DoneIcon) />}
  />
  <WithStyles(Menu)
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "bottom",
      }
    }
    getContentAnchorEl={null}
    id="more-menu"
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
  >
    <WithStyles(MenuItem)
      onClick={[Function]}
    >
      <WithStyles(ListItemIcon)>
        <pure(DeleteIcon) />
      </WithStyles(ListItemIcon)>
      Delete Note
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      onClick={[Function]}
    >
      <WithStyles(ListItemIcon)>
        <pure(ClearIcon) />
      </WithStyles(ListItemIcon)>
      Unpin Note
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      onClick={[Function]}
    >
      <WithStyles(ListItemIcon)>
        <pure(ArchiveIcon) />
      </WithStyles(ListItemIcon)>
      Archive Note
    </WithStyles(MenuItem)>
  </WithStyles(Menu)>
  <ConfirmationDialog
    message="Are you sure?"
    negative="No"
    onNegative={[Function]}
    onPositive={[Function]}
    open={false}
    positive="Yes"
    title="You are about to delete note: undefined"
  />
  <ConfirmationDialog
    message="Are you sure?"
    negative="No"
    onNegative={[Function]}
    onPositive={[Function]}
    open={false}
    positive="Yes"
    title="If you close this editor, you will lose your changes."
  />
  <WithStyles(CardContent)>
    <WithStyles(Dialog)
      classes={
        Object {
          "root": undefined,
        }
      }
      fullScreen={false}
      fullWidth={true}
      maxWidth="lg"
      onClose={[Function]}
      open={false}
    >
      <React.Suspense
        fallback={
          <Loading
            color="#000000"
            delay={0}
            height={64}
            type="spin"
            width={64}
          />
        }
      >
        <lazy
          innerRef={
            Object {
              "current": null,
            }
          }
          note={
            Object {
              "body": "",
              "parent_note_id": undefined,
              "tags": Array [],
              "title": "",
            }
          }
          onSave={[Function]}
        />
      </React.Suspense>
    </WithStyles(Dialog)>
    <WithStyles(Tags) />
    <ReactMarkdown
      astPlugins={
        Array [
          [Function],
        ]
      }
      escapeHtml={false}
      parserOptions={Object {}}
      plugins={Array []}
      rawSourcePos={false}
      renderers={
        Object {
          "text": [Function],
        }
      }
      skipHtml={false}
      sourcePos={false}
      transformLinkUri={[Function]}
    />
    <WithStyles(Grid)
      container={true}
      spacing={8}
    >
      <withRouter(Connect(WithStyles(NoteList)))
        deleteNote={[Function]}
        depth={1}
        notes={Map {}}
        search=""
        updateNote={[Function]}
      />
    </WithStyles(Grid)>
  </WithStyles(CardContent)>
</WithStyles(Card)>
`;

exports[`<Note /> matches snapshot 3`] = `
<WithStyles(Card)
  className="undefined"
>
  <WithStyles(CardHeader)
    action={
      <React.Fragment>
        <WithStyles(IconButton)
          aria-label="Add Subnote"
          onClick={[Function]}
        >
          <pure(LibraryAddIcon) />
        </WithStyles(IconButton)>
        <WithStyles(IconButton)
          aria-label="Edit Note"
          onClick={[Function]}
        >
          <pure(EditIcon) />
        </WithStyles(IconButton)>
        <WithStyles(IconButton)
          aria-label="More Options"
          onClick={[Function]}
        >
          <pure(MoreVertIcon) />
        </WithStyles(IconButton)>
      </React.Fragment>
    }
    avatar={null}
  />
  <WithStyles(Menu)
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "bottom",
      }
    }
    getContentAnchorEl={null}
    id="more-menu"
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
  >
    <WithStyles(MenuItem)
      onClick={[Function]}
    >
      <WithStyles(ListItemIcon)>
        <pure(DeleteIcon) />
      </WithStyles(ListItemIcon)>
      Delete Note
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      onClick={[Function]}
    >
      <WithStyles(ListItemIcon)>
        <pure(DoneIcon) />
      </WithStyles(ListItemIcon)>
      Pin Note
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      onClick={[Function]}
    >
      <WithStyles(ListItemIcon)>
        <pure(UnarchiveIcon) />
      </WithStyles(ListItemIcon)>
      Unarchive Note
    </WithStyles(MenuItem)>
  </WithStyles(Menu)>
  <ConfirmationDialog
    message="Are you sure?"
    negative="No"
    onNegative={[Function]}
    onPositive={[Function]}
    open={false}
    positive="Yes"
    title="You are about to delete note: undefined"
  />
  <ConfirmationDialog
    message="Are you sure?"
    negative="No"
    onNegative={[Function]}
    onPositive={[Function]}
    open={false}
    positive="Yes"
    title="If you close this editor, you will lose your changes."
  />
  <WithStyles(CardContent)>
    <WithStyles(Dialog)
      classes={
        Object {
          "root": undefined,
        }
      }
      fullScreen={false}
      fullWidth={true}
      maxWidth="lg"
      onClose={[Function]}
      open={false}
    >
      <React.Suspense
        fallback={
          <Loading
            color="#000000"
            delay={0}
            height={64}
            type="spin"
            width={64}
          />
        }
      >
        <lazy
          innerRef={
            Object {
              "current": null,
            }
          }
          note={
            Object {
              "body": "",
              "parent_note_id": undefined,
              "tags": Array [],
              "title": "",
            }
          }
          onSave={[Function]}
        />
      </React.Suspense>
    </WithStyles(Dialog)>
    <WithStyles(Tags) />
    <ReactMarkdown
      astPlugins={
        Array [
          [Function],
        ]
      }
      escapeHtml={false}
      parserOptions={Object {}}
      plugins={Array []}
      rawSourcePos={false}
      renderers={
        Object {
          "text": [Function],
        }
      }
      skipHtml={false}
      sourcePos={false}
      transformLinkUri={[Function]}
    />
    <WithStyles(Grid)
      container={true}
      spacing={8}
    >
      <withRouter(Connect(WithStyles(NoteList)))
        deleteNote={[Function]}
        depth={1}
        notes={Map {}}
        search=""
        updateNote={[Function]}
      />
    </WithStyles(Grid)>
  </WithStyles(CardContent)>
</WithStyles(Card)>
`;
